### PR TITLE
Added 4by3 keyboard

### PR DESCRIPTION
**Pull request by eliassjogreen**
## Description

Added the 4by3 keypad, a small macro pad using the pro micro.

Also added a few different keymaps to the keypad:
- arrow_pad
- default
- function_horizontal
- function_vertical_left
- function_vertical_right
- numpad_horizontal
- numpad_vertical_left
- numpad_vertical_right

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes (or in this case the keyboard) and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/4by3/rules.mk [hunk 1]
```diff
@@ -0,0 +1,11 @@
+MCU 				= atmega32u4
+F_CPU 				= 16000000
```
Comment (fauxpark) [inline keyboards/4by3/rules.mk]: As of #6253, `F_CPU`, `F_USB`, `ARCH`, and `OPT_DEFS` can actually be omitted for the ATmega32U4 if they are left at default :)

Comment (eliassjogreen) [reply keyboards/4by3/rules.mk]: Removed `F_CPU`, `F_USB`, `ARCH`, and `OPT_DEFS`

**Comment by noroadsleft** [comment]
> This looks good, for the most part.
> 
> My only issue here, is that we generally denote row by column, not column by row.
> 
> So, it would be 3x4, not 4x3.
> 
> If you look at the default layouts, you'll see this, as well.
> 
> It's a minor thing, but it bugs me.

It bugs me too. Not only that, the image on the readme says "3by4."

**Comment by eliassjogreen** [comment]
> 
> 
> > This looks good, for the most part.
> > My only issue here, is that we generally denote row by column, not column by row.
> > So, it would be 3x4, not 4x3.
> > If you look at the default layouts, you'll see this, as well.
> > It's a minor thing, but it bugs me.
> 
> It bugs me too. Not only that, the image on the readme says "3by4."

The PCB shown in the readme was is only a prototype and the text saying 3by4 was only a typo which has been fixed.

**Comment by eliassjogreen** [comment]
Also it is not possible to have the macro resemble the physical orientation due to the fact you can only have as many rows as defined in config.h without getting the `excess elements in array initializer` error.

**Comment by drashna** [comment]
@eliassjogreen that's true for the physical matrix, but the `LAYOUT` macro allows you to remap that matrix to ANYTHING.  

For instance, the Planck rev6 is a 4x12 ortho board, but the physical matrix is actually 6x8. 

https://github.com/qmk/qmk_firmware/blob/master/keyboards/planck/rev6/rev6.h#L89-L104

So, the bottom half is what the physical matrix is, while the is literally any way you want it.

**Comment by eliassjogreen** [comment]
Thanks! Got it the otherway around when i got the `excess elements in array initializer` but now it works fine :D

File: keyboards/4by3/info.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        "LAYOUT_horizontal": {
+            "key_count": 12,
+            "layout": [
+                { "x": 0, "y": 0 }, { "x": 1, "y": 0 }, { "x": 2, "y": 0 }, { "x": 3, "y": 0 },
+                { "x": 0, "y": 1 }, { "x": 1, "y": 1 }, { "x": 2, "y": 1 }, { "x": 3, "y": 1 },
+                { "x": 0, "y": 2 }, { "x": 1, "y": 2 }, { "x": 2, "y": 2 }, { "x": 3, "y": 2 }
+            ]
+        },
+        "LAYOUT_vertical_right": {
+            "key_count": 12,
```
Comment (noroadsleft) [inline keyboards/4by3/info.json]: You can actually set the `width` and `height` properties per layout:

```suggestion
            "key_count": 12,
            "width": 3,
            "height": 4,
```

Reference: [info.json Format - QMK Firmware Documentation](https://docs.qmk.fm/#/reference_info_json)

File: keyboards/4by3/info.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            "layout": [
+                { "x": 0, "y": 2 }, { "x": 0, "y": 1 }, { "x": 0, "y": 0 },
```
Comment (noroadsleft) [inline keyboards/4by3/info.json]: Lines 19-22 should actually be:

```json
                { "x": 0, "y": 0 }, { "x": 1, "y": 0 }, { "x": 2, "y": 0 },
                { "x": 0, "y": 1 }, { "x": 1, "y": 1 }, { "x": 2, "y": 1 },
                { "x": 0, "y": 2 }, { "x": 1, "y": 2 }, { "x": 2, "y": 2 },
                { "x": 0, "y": 3 }, { "x": 1, "y": 3 }, { "x": 2, "y": 3 }
```

The layout macro takes care of mapping the physical location to the electrical position, but it doesn't provide any information about key sizes or positions, which is what this does.

Your current implementation would force QMK Configurator users to map the keys as follows to get the `arrowpad` keymap:

image

... which can be misleading.

File: keyboards/4by3/info.json [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+                { "x": 1, "y": 2 }, { "x": 1, "y": 1 }, { "x": 1, "y": 0 },
+                { "x": 2, "y": 2 }, { "x": 2, "y": 1 }, { "x": 2, "y": 0 },
+                { "x": 3, "y": 2 }, { "x": 3, "y": 1 }, { "x": 3, "y": 0 }
+            ]
+        },
+        "LAYOUT_vertical_left": {
+            "key_count": 12,
```
Comment (noroadsleft) [inline keyboards/4by3/info.json]:
```suggestion
            "key_count": 12,
            "width": 3,
            "height": 4,
```

File: keyboards/4by3/info.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            "layout": [
+                { "x": 3, "y": 0 }, { "x": 3, "y": 1 }, { "x": 3, "y": 2 },
```
Comment (noroadsleft) [inline keyboards/4by3/info.json]: Lines 28-31 should actually be:

```json
                { "x": 0, "y": 0 }, { "x": 1, "y": 0 }, { "x": 2, "y": 0 },
                { "x": 0, "y": 1 }, { "x": 1, "y": 1 }, { "x": 2, "y": 1 },
                { "x": 0, "y": 2 }, { "x": 1, "y": 2 }, { "x": 2, "y": 2 },
                { "x": 0, "y": 3 }, { "x": 1, "y": 3 }, { "x": 2, "y": 3 }
```

File: keyboards/4by3/readme.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
+# 4by3
+
+![4by3](https://i.imgur.com/Ykb7evL.jpg)
+
+A 12 key mechanical keypad.
+
+Keyboard Maintainer: [eliassjogreen](https://github.com/eliassjogreen)  
+Hardware Supported: 3by4
```
Comment (noroadsleft) [inline keyboards/4by3/readme.md]:
```suggestion
Hardware Supported: 4by3  
```

File: keyboards/4by3/rules.mk [hunk 7]
```diff
@@ -0,0 +1,9 @@
+MCU 				= atmega32u4
+BOOTLOADER 			= caterina
+
+EXTRAKEY_ENABLE 	= yes
+NKRO_ENABLE 		= yes
+CONSOLE_ENABLE		= yes
+COMMAND_ENABLE		= yes
+
+LAYOUTS				= horizontal vertical_right vertical_left
```
Comment (noroadsleft) [inline keyboards/4by3/rules.mk]: The `LAYOUTS` rule is only for the [Community Layouts feature](https://docs.qmk.fm/#/feature_layouts). As none of the listed options are currently valid choices, this line should be deleted.

```suggestion
```

File: keyboards/4by3/4by3.h [hunk 8]
```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include "quantum.h"
+
+#define LAYOUT_horizontal( \
```
Comment (noroadsleft) [inline keyboards/4by3/4by3.h]: Not a hard requirement, but I think some block comments here regarding PCB/Pro Micro orientation in this file would help explain what's going on for anyone who may look in this file in future.

Something like this above each macro definition would be good:

```h
/* LAYOUT_horizontal
 *             ┌───┐
 *             │USB│
 * ┌───┬───┬───┼───┤
 * │K00│K01│K02│K03│
 * ├───┼───┼───┼───┤
 * │K10│K11│K12│K13│
 * ├───┼───┼───┼───┤
 * │K20│K21│K22│K23│
 * └───┴───┴───┴───┘
 */
```

File: keyboards/4by3/4by3.h [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    K21, K11, K01, \
+    K22, K12, K02, \
+    K23, K13, K03  \
+) { \
+	{ K00, K01, K02, K03 }, \
+	{ K10, K11, K12, K13 }, \
+	{ K20, K21, K22, K23 }  \
+}
+
+/* LAYOUT_vertical_right
```
Comment (fauxpark) [inline keyboards/4by3/4by3.h]:
```suggestion
/* LAYOUT_vertical_left
```

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**